### DISCUSSION TITLE
Add better install instructions for diesel_cli

### DISCUSSION THREAD

**Discussion by RKochenderfer** [diesel-rs/diesel#2971] [Ideas]
# The Issue
A common issue when developing on Linux for installing diesel_cli is getting 
```
error: linking with `cc` failed: exit status: 1
  |
  = note: "cc" "-m64" "/tmp/cargo-installw2sZxC/release/deps/diesel-a512bb13f6c51e5e.diesel.52123629-cgu.0.rcgu.o" "/tmp/cargo-installw2sZxC/release/deps/diesel-a512bb13f6c51e5e.diesel.52123629-cgu.1.rcgu.o" "/tmp/cargo-installw2sZxC/release/deps/diesel-a512bb13f6c51e5e.diesel.52123629-cgu.10.rcgu.o" "/tmp/cargo-installw2sZxC/release/deps/diesel-a512bb13f6c51e5e.diesel.52123629-cgu.11.rcgu.o" "/tmp/cargo-installw2sZxC/release/deps/diesel-a512bb13f6c51e5e.diesel.52123629-cgu.12.rcgu.o" "/tmp/cargo-installw2sZxC/release/deps/diesel-a512bb13f6c51e5e.diesel.52123629-cgu.13.rcgu.o" "/tmp/cargo-installw2sZxC/release/deps/diesel-a512bb13f6c51e5e.diesel.52123629-cgu.14.rcgu.o" "/tmp/cargo-installw2sZxC/release/deps/diesel-a512bb13f6c51e5e.diesel.52123629-cgu.15.rcgu.o" "/tmp/cargo-installw2sZxC/release/deps/diesel-a512bb13f6c51e5e.diesel.52123629-cgu.2.rcgu.o" "/tmp/cargo-installw2sZxC/release/deps/diesel-a512bb13f6c51e5e.diesel.52123629-cgu.3.rcgu.o" "/tmp/cargo-installw2sZxC/release/deps/diesel-a512bb13f6c51e5e.diesel.52123629-cgu.4.rcgu.o" "/tmp/cargo-installw2sZxC/release/deps/diesel-a512bb13f6c51e5e.diesel.52123629-cgu.5.rcgu.o" "/tmp/cargo-installw2sZxC/release/deps/diesel-a512bb13f6c51e5e.diesel.52123629-cgu.6.rcgu.o" "/tmp/cargo-installw2sZxC/release/deps/diesel-a512bb13f6c51e5e.diesel.52123629-cgu.7.rcgu.o" "/tmp/cargo-installw2sZxC/release/deps/diesel-a512bb13f6c51e5e.diesel.52123629-cgu.8.rcgu.o" "/tmp/cargo-installw2sZxC/release/deps/diesel-a512bb13f6c51e5e.diesel.52123629-cgu.9.rcgu.o" "/tmp/cargo-installw2sZxC/release/deps/diesel-a512bb13f6c51e5e.21y6vcjz30orzgu2.rcgu.o" "-Wl,--as-needed" "-L" "/tmp/cargo-installw2sZxC/release/deps" "-L" "/home/rkochenderfer/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-Wl,-Bstatic" "/tmp/cargo-installw2sZxC/release/deps/libtoml-83b13cbf0d12d2cc.rlib" "/tmp/cargo-installw2sZxC/release/deps/libtempfile-41885ea3d41980e2.rlib" "/tmp/cargo-installw2sZxC/release/deps/librand-cd35038c04508d62.rlib" "/tmp/cargo-installw2sZxC/release/deps/librand_chacha-896f4e3c3c2b6800.rlib" "/tmp/cargo-installw2sZxC/release/deps/libppv_lite86-b4dbb70e1e0a5c67.rlib" "/tmp/cargo-installw2sZxC/release/deps/librand_core-6731228e1961cbd3.rlib" "/tmp/cargo-installw2sZxC/release/deps/libgetrandom-c7bdb720a82b98b7.rlib" "/tmp/cargo-installw2sZxC/release/deps/libremove_dir_all-fb73e0bec459429d.rlib" "/tmp/cargo-installw2sZxC/release/deps/libserde-fa20d0743b0f13a9.rlib" "/tmp/cargo-installw2sZxC/release/deps/libmigrations_internals-729578a3353c0164.rlib" "/tmp/cargo-installw2sZxC/release/deps/libdotenv-508bd7839b83c585.rlib" "/tmp/cargo-installw2sZxC/release/deps/libregex-4051bf3ccb56d7b2.rlib" "/tmp/cargo-installw2sZxC/release/deps/libutf8_ranges-aba36ac13c732a2b.rlib" "/tmp/cargo-installw2sZxC/release/deps/libregex_syntax-de24eff880ae7e88.rlib" "/tmp/cargo-installw2sZxC/release/deps/libucd_util-eb26796c30bb7a7c.rlib" "/tmp/cargo-installw2sZxC/release/deps/libthread_local-63ef90da7c2bacb9.rlib" "/tmp/cargo-installw2sZxC/release/deps/liblazy_static-0df5d4f37a836555.rlib" "/tmp/cargo-installw2sZxC/release/deps/libaho_corasick-7c6063df01d3188f.rlib" "/tmp/cargo-installw2sZxC/release/deps/liberror_chain-4a12c0accc8e2ee3.rlib" "/tmp/cargo-installw2sZxC/release/deps/libbacktrace-11376978ca7e6b93.rlib" "/tmp/cargo-installw2sZxC/release/deps/libminiz_oxide-64bbc3dfcc85a6b0.rlib" "/tmp/cargo-installw2sZxC/release/deps/libadler-56293137e09c9368.rlib" "/tmp/cargo-installw2sZxC/release/deps/libobject-1e13e5b4501c12d2.rlib" "/tmp/cargo-installw2sZxC/release/deps/libmemchr-06d76bcb96a352b9.rlib" "/tmp/cargo-installw2sZxC/release/deps/libaddr2line-fa7a28a8be6201a7.rlib" "/tmp/cargo-installw2sZxC/release/deps/libgimli-c3460830815a04ac.rlib" "/tmp/cargo-installw2sZxC/release/deps/libcfg_if-a04351a1e37ac403.rlib" "/tmp/cargo-installw2sZxC/release/deps/librustc_demangle-742846db1c8a4c60.rlib" "/tmp/cargo-installw2sZxC/release/deps/libdiesel-9bd433becfa5cf3e.rlib" "/tmp/cargo-installw2sZxC/release/deps/liblibsqlite3_sys-3399c1bbb69fa7aa.rlib" "/tmp/cargo-installw2sZxC/release/deps/libbyteorder-20fda3f2a9be6fa6.rlib" "/tmp/cargo-installw2sZxC/release/deps/libclap-cc3b326180431dd1.rlib" "/tmp/cargo-installw2sZxC/release/deps/libvec_map-5d3b872e0a1408b5.rlib" "/tmp/cargo-installw2sZxC/release/deps/libtextwrap-c273469f9debde5a.rlib" "/tmp/cargo-installw2sZxC/release/deps/libunicode_width-b88877dbb8531487.rlib" "/tmp/cargo-installw2sZxC/release/deps/libstrsim-65d231e2011670e5.rlib" "/tmp/cargo-installw2sZxC/release/deps/libbitflags-045e5e9961c5ccd8.rlib" "/tmp/cargo-installw2sZxC/release/deps/libatty-0ef09533989378ef.rlib" "/tmp/cargo-installw2sZxC/release/deps/libansi_term-594c980194a3606c.rlib" "/tmp/cargo-installw2sZxC/release/deps/libchrono-954472fd115ee85c.rlib" "/tmp/cargo-installw2sZxC/release/deps/libnum_integer-e29a57430a02535e.rlib" "/tmp/cargo-installw2sZxC/release/deps/libnum_traits-68fa78ce62b4eea2.rlib" "/tmp/cargo-installw2sZxC/release/deps/libtime-9f9dcc8c1366395e.rlib" "/tmp/cargo-installw2sZxC/release/deps/liblibc-8f81e38aa17dea8d.rlib" "-Wl,--start-group" "/home/rkochenderfer/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd-7c582493123fc1dd.rlib" "/home/rkochenderfer/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libpanic_unwind-1392776590706175.rlib" "/home/rkochenderfer/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libminiz_oxide-2eb6edf4d031cd1e.rlib" "/home/rkochenderfer/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libadler-33a7ad3b5f7fedf6.rlib" "/home/rkochenderfer/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libobject-7d32adce541987d9.rlib" "/home/rkochenderfer/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libmemchr-5cb369120f224726.rlib" "/home/rkochenderfer/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libaddr2line-6ab2efd1d2f431a9.rlib" "/home/rkochenderfer/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libgimli-567f611439253c7e.rlib" "/home/rkochenderfer/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd_detect-ebbc63efd6d2efc5.rlib" "/home/rkochenderfer/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_demangle-862830f0d224a2e1.rlib" "/home/rkochenderfer/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libhashbrown-cad0401ae7a80e32.rlib" "/home/rkochenderfer/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_std_workspace_alloc-fd54290077194763.rlib" "/home/rkochenderfer/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libunwind-4fc3313c8ccb1ec0.rlib" "/home/rkochenderfer/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcfg_if-19bf8dffe82b09d4.rlib" "/home/rkochenderfer/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/liblibc-afc95b1640c4beca.rlib" "/home/rkochenderfer/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/liballoc-aff6658baa87e3d1.rlib" "/home/rkochenderfer/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_std_workspace_core-8be8a1689a4f7b48.rlib" "/home/rkochenderfer/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcore-5284934f66073844.rlib" "-Wl,--end-group" "/home/rkochenderfer/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcompiler_builtins-2a0b2a4f96acb821.rlib" "-Wl,-Bdynamic" "-lsqlite3" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc" "-Wl,--eh-frame-hdr" "-Wl,-znoexecstack" "-L" "/home/rkochenderfer/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-o" "/tmp/cargo-installw2sZxC/release/deps/diesel-a512bb13f6c51e5e" "-Wl,--gc-sections" "-pie" "-Wl,-zrelro" "-Wl,-znow" "-Wl,-O1" "-nodefaultlibs"
  = note: /usr/bin/ld: cannot find -lsqlite3
          collect2: error: ld returned 1 exit status
          

error: failed to compile `diesel_cli v1.4.1`, intermediate artifacts can be found at `/tmp/cargo-installw2sZxC`

Caused by:
  could not compile `diesel_cli` due to previous error
```

This is such a common error with an unhelpful error that has such an easy fix that I believe it needs to be added to the official documentation as a prerequisite before installing the diesel_cli. In the getting started documentation's current state it makes no mention of  that `libpq-dev` is required to successfully compile the suggested command on Linux on the get started page for the postgres only install that is, `cargo install diesel_cli --no-default-features --features postgres`.

The current instruction on the page leads potential users to believe that the issue is that they do not have the database installed. However installing PostgreSQL or Sqlite3 through the suggested methods does *not* install the necessary libraries that are needed to compile the diesel_cli.

# Solution
Add a common install errors section to the official diesl.rs site or add an install assistance section to the official diesel.rs site. This would be the easiest way to help new users of the library quickly get to using it and limiting blocking factors

**Comment by weiznich** [comment]
This was already discussed multiple times. Unfortunately it's not as simple as writing down that you need to install `libpq-dev`, as each linux distribution names that package a bit different and that does windows…. At least I cannot maintain a list of all operating systems and I do not want to have a small list of "popular" systems as this would sound like "those" other distributions/systems are not supported, which is not true.

That written: I would accept a PR to the [web page repository](https://github.com/sgrif/diesel.rs-website/blob/eebab3fb997f5fd75cf4e31fb4a35b094ba3e0b5/src/guides/getting-started.md?plain=1#L93-L102) to update the linked statement to say something more specific in that way that the postgres feature requires `libpq`, the mysql feature requires `libmysqlclient` and the `sqlite` feature requires `libsqlite3` (or the bundled flag) and that you need to check the documentation of your operating system/those libraries for ways to install those dependencies.
I also would welcome work to [`pq-sys`](https://github.com/sgrif/pq-sys) or [`mysqlclient-sys`](https://github.com/sgrif/mysqlclient-sys) to add a `bundled` feature flag there as well, so that those crates may build and statically link the corresponding dependencies on their own.

**Reply by Mingun** [reply]
> as this would sound like "those" other distributions/systems are not supported

I think this is only you wrong feeling

**Reply by weiznich** [reply]
@Mingun Ok to word that differently: I personally do not want to maintain such a list as I do not have the motivation to do that. I would be completely happy if someone volunteers to do that. As soon as I have the necessary trust that this would be a long term commitment to maintain such list I'm happily share the corresponding rights with that contributor and from that point they are free to include this information in the documentation. For me personally there are other more important things to spend my time on.

**Reply by weiznich** [reply]
Also this is already such a guide/list available in more than one place. See [here](https://github.com/diesel-rs/diesel/blob/master/guide_drafts/backend_installation.md) and [here](https://github.com/diesel-rs/diesel/tree/master/diesel_cli#installation) for examples. I feel that people keep ignoring/missing those so I do not see any evidence that adding those information in even more locations will help in any way.

**Reply by Mingun** [reply]
At least the second link is slightly incorrect which could be possible source of mistakes. It requires to install `libpq`, but
1. This is not a name of package, but _it looks like_ a package name on linux
2. The correct package in the very popular .deb format should be `libpq-dev`
3. The other variants not suggesting something like `libopenssl`, `libmysql` or `libsqlite`, so people think that `libpq` is an actual package name, especially if they are aware that database is called Postgres

So I think that people following the guide as it written and tried to install `libpq` or analogues which does not contains development files.

I would suggest to change `libpq` to `libpq-dev` or to `postgres`.

Anyway, if this is the only place of this documentation (both your links points to GitHub, not to the diesel.rs site) it cannot be called "highly available". If not, well, even you firstly points to this place, so it can tell something. But ordinary developer starts learning from the diesel.rs site, not from reading code/readme on GitHub.

**Reply by weiznich** [reply]
@Mingun 

> This is not a name of package, but it looks like a package name on linux
> 1. The correct package in the very popular .deb format should be libpq-dev

If we change that to `libpq-dev` people will complain that fedora does not have a `libpq-dev` package, that's the reason why I don't want to list any package name at all. Also arguing with "popularity" here is not meaningful, as this can change and there are other "very popular" distributions with other package names.

> The other variants not suggesting something like libopenssl, libmysql or libsqlite, so people think that libpq is an actual package name, especially if they are aware that database is called Postgres

Feel free to submit a PR changing that.

> But ordinary developer starts learning from the diesel.rs site, not from reading code/readme on GitHub.

At least based on the question some people ask they don't spend time to read any guide on the web page, so adding the information there will also not help.

> Anyway, if this is the only place of this documentation (both your links points to GitHub, not to the diesel.rs site) it cannot be called "highly available". If not, well, even you firstly points to this place, so it can tell something.

Feel free to submit a PR changing that or stop arguing about it. At least for me that's something that just does not have priority. This is my last answer to you about this topic. You seems to be just a person that always complains but never contributes something. That's probably not the best way to convince someone from changing something.

**Reply by RKochenderfer** [reply]
I created a pull request to the get_started documentation with an explicit link to the backend installation instructions that should help people identify that they need to have this installed and also guide them into how and what needs to be installed [here](https://github.com/sgrif/diesel.rs-website/pull/144)
